File: dart/function.dart
main() {
  // * lambda;
  num bar(num a, num b) => a + b;

  const list = <int>[1, 2];

  final _ = list.fold(0, bar);

  // * function
  void foo(num a, num b) {
    print(a + b);
  }

  // * apply
  Function.apply(foo, [1, 2]);

  // *  特殊操作
  (() => () {})();

  // * callable class
  final foo1 = Foo();

  foo1()()().foo()().foo();

  // * async function
  wait() async {
    await Future.delayed(const Duration(seconds: 1));
    print('done');
  }

  wait();

  Iterable<int> iter() sync* {
    int count = 0;
    while (true) {
      yield count++;
    }
  }

  final v = iter().take(1);
  print(v);
}

class Foo {
  get foo => this;
  call() => this;
}

File: dart/function.js
// * lambda
const sum = (acc, e) => acc + e;

[1, 2].reduce(sum, 0);

// * function
const foo = (a, b) => {
	console.log(a + b);
};

// * call apply
const arg = [1, 2];
foo.call(null, ...arg);
foo.apply(null, arg);

// * 特殊操作
(() => () => {})();

// * 函数也是对象
const bar = () => bar;
bar.bar = bar;

bar()()()
	.bar()()
	.bar();

// * async
const wait = async () => {
	await new Promise(res => setTimeout(res, 1000));
	console.log('done');
};

// * generator
function* produce() {
	let count = 0;
	while (true) {
		yield count++;
	}
}
const a = produce();
const { value } = a.next();
